airbnb_project/listings/harvester_app/harvester/spiders/listings_spider.py

```diff
@@ -0,0 +1,312 @@
+                longitude = coordinate.get("longitude", "")
+                room_type = listing.get("roomTypeCategory", "")
+                page_url = API_CALL.format(n(t("StayListing", listing_id)))
+                airbnb_params = {
+                    "airbnb_listing_id": listing_id,
+                    "title": title,
+                    "name": name,
+                    "latitude": latitude,
+                    "longitude": longitude,
+                    "room_type": room_type
+                }
+
+                # A second request is made to get the registration number and other information using the listing_id
+                # which is scarped
+                if listing_id and listing_id not in ListingsSpider.ALREADY_SCRAPED_LISTINGS:
+                    yield Request(url=page_url, callback=self.handle_listing,
+                                  headers={'X-Airbnb-Api-Key': API_KEY},
+                                  meta={'airbnb_params': airbnb_params})
+                else:
+                    print("duplicate", listing_id)
+
+            except Exception as e:
+                print("error_listing", result)
+                print("exception", e.args[0])
+
+        if len(ListingsSpider.next_page_cursors) != 0:
+            cursor_id = ListingsSpider.next_page_cursors.pop()
+            next_url = f'{current_url}&cursor={cursor_id}'
+            yield response.follow(next_url, callback=self.parse)
+
+    @staticmethod
+    def handle_listing(response):
+        """
+        Parse method for extracting details from an Airbnb listing page.
+
+        Args:
+            response (Response): The response object from the request.
+
+         Yields:
+            AirBnBListingItem: An AirBnBListingItem object containing details of an Airbnb listing.
+        """
+        airbnb_params = response.meta.get('airbnb_params', {})
+        listing_item = ExpandedAirBnBListingItem(
+            airbnb_listing_id=airbnb_params.get('airbnb_listing_id'),
+            title=airbnb_params.get('title'),
+            name=airbnb_params.get('name'),
+            latitude=airbnb_params.get('latitude'),
+            longitude=airbnb_params.get('longitude'),
+            room_type=airbnb_params.get('room_type')
+        )
+
+        script_tag_json = json.loads(response.text)
+
+        location = ""
+        person_capacity = ""
+        try:
+            metadata = ListingsSpider._get_metadata_from_json(script_tag_json)
+            if metadata is not None:
+                sharing_config = metadata.get("sharingConfig", {})
+                location = sharing_config.get("location", "")
+                person_capacity = sharing_config.get("personCapacity", "")
+
+        except Exception as e:
+            print(e)
+        listing_item['location'] = location
+        listing_item['person_capacity'] = person_capacity
+        registration_number = ""
+        number_of_beds = ""
+        number_of_baths = ""
+        try:
```

Comment:
```airbnb_params = response.meta.get('airbnb_params', {})
        listing_item = ExpandedAirBnBListingItem(
            airbnb_listing_id=airbnb_params.get('airbnb_listing_id'),
            title=airbnb_params.get('title'),
            name=airbnb_params.get('name'),
            latitude=airbnb_params.get('latitude'),
            longitude=airbnb_params.get('longitude'),
            room_type=airbnb_params.get('room_type')
        )
        try:
            script_tag_json = json.loads(response.text)
            ListingsSpider._parse_capacity_and_location(script_tag_json, listing_item)
            ListingsSpider._parse_listings_number(script_tag_json, listing_item)
        except Exception as e:
            print(e)
        finally:
            yield listing_item
```
Would something like this work? I moved the code the separate functions that have their own try and catch blocks.

Reply:
Absolutely. This comes down to a design preference. I'll leave it up to you but the bottom line here is, if each function has their own specific exception then that's ok. Otherwise, we can probably use 1 try/except and call all functions back to back.

For example, if functions have different exceptions:
`
func_a():
    try:
        ...
    except TypeError as e:
        ...
func_b():
    try:
        ...
    except IndexError as e:
        ...
func_c():
    try:
        ...
    except AttributeError as e:
        ...
`

Or, the other way around:
`
try:
    func_a()
    func_b()
    func_c()
except Exception as e:
        ...
`

Does this make sense?

Reply:
yeah, makes sense.